### ISSUE TITLE
Using Elephas + Keras to build a Neural Collaborative Filter : TypeError: cannot pickle '_thread.lock' object 

### ISSUE THREAD

**Issue by yvsandeep** [danielenricocahall/elephas#30]
**Describe the bug**
I am trying to build a Neural collaborative filtering on million song data approach using Keras and distributing it over a spark cluster using Elephas. Data consists of triplets( User ID, Song ID, Number of Plays). I was able to run the model using pandas+keras+tensorflow but facing issues when I use it with elephas in pyspark. When i call the 
`spark_model.fit(rdd_final, epochs=5, batch_size=32, verbose=0, validation_split=0.1,categorical=False, nb_classes=None)`

I get the error: 
>>> Fit model
Traceback (most recent call last):
  File "/Users/sandeep/Desktop/BUCourses/CS777/Term_Project/elephas_ncf.py", line 58, in <module>
    spark_model.fit(rdd_final, epochs=5, batch_size=32, verbose=0, validation_split=0.1,categorical=False, nb_classes=None)
  File "/Users/sandeep/anaconda3/lib/python3.11/site-packages/elephas/spark_model.py", line 186, in fit
    self._fit(rdd, **kwargs)
  File "/Users/sandeep/anaconda3/lib/python3.11/site-packages/elephas/spark_model.py", line 197, in _fit
    self.start_server()
  File "/Users/sandeep/anaconda3/lib/python3.11/site-packages/elephas/spark_model.py", line 149, in start_server
    self.parameter_server.start()
  File "/Users/sandeep/anaconda3/lib/python3.11/site-packages/elephas/parameter/server.py", line 83, in start
    self.server.start()
  File "/Users/sandeep/anaconda3/lib/python3.11/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/Users/sandeep/anaconda3/lib/python3.11/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sandeep/anaconda3/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/sandeep/anaconda3/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/sandeep/anaconda3/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/sandeep/anaconda3/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/sandeep/anaconda3/lib/python3.11/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle '_thread.lock' object
 
**To Reproduce**

`
from pyspark.sql import SparkSession
from pyspark.ml.feature import StringIndexer, VectorAssembler
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Embedding, Flatten, Dot
from tensorflow.keras import optimizers
from elephas.ml_model import ElephasEstimator
from pyspark import SparkContext, SparkConf
from tensorflow.keras.optimizers import SGD
from elephas.spark_model import SparkModel

def row_to_tuple(row):
    # 'user_id_encoded' and 'track_id_encoded' are feature columns
    # 'frequency' is the label column
    return ([row.user_id_encoded, row.track_id_encoded], row.frequency)

if __name__ == "__main__":
    conf = SparkConf().setAppName('Elephas_App').setMaster('local[8]')
    sc = SparkContext(conf=conf)
    sc.stop()
    spark = SparkSession.builder.appName("NeuralCollaborativeFiltering").getOrCreate()

    # Read and preprocess data
    df = spark.read.csv('train_triplets.txt', sep='\t', header=False, inferSchema=True)
    df = df.withColumnRenamed('_c0', 'user_id').withColumnRenamed('_c1', 'track_id').withColumnRenamed('_c2', 'frequency')
    df = df.limit(1000)

    indexer_user = StringIndexer(inputCol="user_id", outputCol="user_id_encoded")
    df = indexer_user.fit(df).transform(df)
    indexer_track = StringIndexer(inputCol="track_id", outputCol="track_id_encoded")
    df = indexer_track.fit(df).transform(df)

    assembler = VectorAssembler(inputCols=["user_id_encoded", "track_id_encoded"], outputCol="features")
    df_transformed = assembler.transform(df)

    # Keras model inputs
    n_users = df.select('user_id').distinct().count()
    n_items = df.select('track_id').distinct().count()
    n_latent_factors = 50  # Example

    user_input = Input(shape=(1,), name='user_input')
    item_input = Input(shape=(1,), name='item_input')
    user_embedding = Embedding(output_dim=n_latent_factors, input_dim=n_users)(user_input)
    item_embedding = Embedding(output_dim=n_latent_factors, input_dim=n_items)(item_input)
    user_vec = Flatten()(user_embedding)
    item_vec = Flatten()(item_embedding)
    dot = Dot(axes=1)([user_vec, item_vec])
    model = Model([user_input, item_input], dot)
    model.compile(optimizer=SGD(), loss='mean_squared_error')
    print(model.summary())

    rdd_final = df_transformed.rdd.map(row_to_tuple)
    print(rdd_final.take(3))

    spark_model = SparkModel(model, frequency='epoch', mode='asynchronous')
    spark_model.fit(rdd_final, epochs=5, batch_size=32, verbose=0, validation_split=0.1,categorical=False, nb_classes=None)

`
**Expected behavior**
fitted model that can be used to generate recommendations. 

**Desktop (please complete the following information):**
 - OS: macOS Sonoma 14.1
 - Python : 3.11.3
 - Tensorflow: 2.13.0
 - Keras : 2.13.1
 - Elephas: 5.0.0
 - Pyspark: 3.4.1

**Additional context**
I am not sure if I can use non-sequential models in Elephas.

**Comment by danielenricocahall** [comment]
Hi @yvsandeep ! Do you have some sample data?

**Comment by yvsandeep** [comment]
Hi @danielenricocahall 
Please find the same below. I have 10000 rows in the same. 
[sample_data.csv](https://github.com/danielenricocahall/elephas/files/13520249/sample_data.csv)

**Comment by danielenricocahall** [comment]
Hello @yvsandeep! I hope all is going well. Could you try changing `from tensorflow.keras.optimizers import SGD` to `from tensorflow.keras.legacy import SGD`? I used your example data and code to reproduce the error, and noticed the import - in Python 3.11+, Elephas currently requires the optimizer import to be from legacy (detailed [here](https://github.com/danielenricocahall/elephas/releases/tag/5.0.0))

**Comment by danielenricocahall** [comment]
Hi @yvsandeep , I hope all is going well. I was able to get your model training with the following snippet:

```python

from keras import Model
from pyspark.ml.feature import StringIndexer, VectorAssembler
from tensorflow.keras.optimizers.legacy import SGD
from tensorflow.keras import Input
from tensorflow.keras.layers import Embedding, Flatten, Dot

...

    def row_to_tuple(row):
        return ([row.user_id_encoded, row.track_id_encoded], row.frequency)

    # Read and preprocess data
    df = spark_session.read.csv('sample_data.csv', header=True, inferSchema=True)
    df = df.limit(100)

    indexer_user = StringIndexer(inputCol="user_id", outputCol="user_id_encoded")
    df = indexer_user.fit(df).transform(df)
    indexer_track = StringIndexer(inputCol="track_id", outputCol="track_id_encoded")
    df = indexer_track.fit(df).transform(df)

    assembler = VectorAssembler(inputCols=["user_id_encoded", "track_id_encoded"], outputCol="features")
    df_transformed = assembler.transform(df)

    # Keras model inputs
    n_users = df.select('user_id').distinct().count()
    n_items = df.select('track_id').distinct().count()
    n_latent_factors = 50  # Example

    user_input = Input(shape=(1,), name='user_input')
    item_input = Input(shape=(1,), name='item_input')
    user_embedding = Embedding(output_dim=n_latent_factors, input_dim=n_users)(user_input)
    item_embedding = Embedding(output_dim=n_latent_factors, input_dim=n_items)(item_input)
    user_vec = Flatten()(user_embedding)
    item_vec = Flatten()(item_embedding)
    dot = Dot(axes=1)([user_vec, item_vec])
    model = Model([user_input, item_input], dot)
    model.compile(optimizer=SGD(), loss='mean_squared_error')
    print(model.summary())

    rdd_final = df_transformed.rdd.map(row_to_tuple)
    print(rdd_final.take(3))

    spark_model = SparkModel(model, frequency='epoch', mode='asynchronous')
    spark_model.fit(rdd_final, epochs=5, batch_size=32, verbose=0, validation_split=0.1)
```

Notice the import change with optimizers. However, it will fail with the following:

```python
E                       ValueError: Layer "model" expects 2 input(s), but it received 1 input tensors. Inputs received: [<tf.Tensor 'IteratorGetNext:0' shape=(None, 2) dtype=float32>]

```

This is because currently, Elephas does not support multi-input models. However, with your sample data, I was able to add the logic required for supporting multiple inputs. I will open a PR soon with this change to review and consider further.

**Comment by yvsandeep** [comment]
Thanks Daniel for the fix. It works great and appreciate the same.

**Comment by danielenricocahall** [comment]
Hi @yvsandeep ! I just released 5.1.0 which includes support for multiple input/multiple output models.

**Comment by danielenricocahall** [comment]
Going to close this out for now, but feel free to open up a new issue if you run into any issues!
